string.c

```diff
@@ -1874,6 +1874,114 @@ rb_str_init(int argc, VALUE *argv, VALUE str)
     return str;
 }
 
+static size_t
+str_embed_size(long capa)
+{
+    return offsetof(struct RString, as.embed.ary) + (sizeof(char) * capa);
+}
+
+static bool
+str_embeddable_p(long capa)
+{
+    return rb_gc_size_allocatable_p(str_embed_size(capa));
+}
+
+/* :nodoc: */
+static VALUE
+rb_str_s_new(int argc, VALUE *argv, VALUE klass)
+{
+    if (klass != rb_cString) {
+        return rb_class_new_instance_pass_kw(argc, argv, klass);
+    }
+
+    static ID keyword_ids[2];
+    VALUE orig, opt, encoding = Qnil, capacity = Qnil;
+    VALUE kwargs[2];
+    rb_encoding *enc = NULL;
+
+    int n = rb_scan_args(argc, argv, "01:", &orig, &opt);
+    if (NIL_P(opt)) {
+        return rb_class_new_instance_pass_kw(argc, argv, klass);
+    }
+
+    keyword_ids[0] = rb_id_encoding();
+    CONST_ID(keyword_ids[1], "capacity");
+    rb_get_kwargs(opt, keyword_ids, 0, 2, kwargs);
+    encoding = kwargs[0];
+    capacity = kwargs[1];
+
+    long termlen = 1;
+
+    if (n == 1) {
+        orig = StringValue(orig);
+    }
+    else {
+        orig = Qnil;
+    }
+
+    if (UNDEF_P(encoding)) {
+        if (!NIL_P(orig)) {
+            encoding = rb_obj_encoding(orig);
+        }
+    }
+
+    if (!UNDEF_P(encoding)) {
+        enc = rb_to_encoding(encoding);
+        termlen = rb_enc_mbminlen(enc);
+    }
+
+    VALUE str;
+
+    // If capacity is nil, we're basically just duping `orig`.
+    if (UNDEF_P(capacity)) {
+        if (NIL_P(orig)) {
+            str = str_new(klass, "", 0);
+            if (enc) {
+                rb_enc_associate(str, enc);
+            }
+            return str;
+        }
+        VALUE copy = str_duplicate(klass, orig);
+        str_modifiable(copy);
```

Comment:
I don't think we need this line? `str_modifiable` just checks that `copy` is not locked nor frozen, both of which should be true since it's newly created.